src/ppx_deriving.ml

```diff
@@ -304,3 +305,181 @@ let extract_typename_of_type_group deriver ~allow_shadowing type_list =
       StringSet.add ty.ptype_name.txt acc in
   List.fold_left add_name StringSet.empty type_list
 
+let dynlink ?(loc=Location.none) filename =
+  let filename = Dynlink.adapt_filename filename in
+  try
+    Dynlink.loadfile filename
+  with Dynlink.Error error ->
+    raise_errorf ~loc "Cannot load %s: %s" filename (Dynlink.error_message error)
+
+let get_plugins () =
+  match Ast_mapper.get_cookie "ppx_deriving" with
+  | Some { pexp_desc = Pexp_tuple exprs } ->
+    exprs |> List.map (fun expr ->
+      match expr with
+      | { pexp_desc = Pexp_constant (Const_string (file, None)) } -> file
+      | _ -> assert false)
+  | Some _ -> assert false
+  | None -> []
+
+let add_plugins plugins =
+  let loaded  = get_plugins () in
+  let plugins = List.filter (fun file -> not (List.mem file loaded)) plugins in
+  List.iter dynlink plugins;
+  let loaded  = loaded @ plugins in
+  Ast_mapper.set_cookie "ppx_deriving"
+    (Exp.tuple (List.map (fun file -> Exp.constant (Const_string (file, None))) loaded))
```

Comment:
Are you sure?  There is a match on @@@findlib.ppxopt in the `structure` local function which calls `add_plugins` and which calls `dynlink`.  I guess it's there to support the `#ppx_deriving` directives in the toplevel.